lib/jara/releaser.rb

```diff
@@ -165,12 +182,77 @@ def exec(command)
           raise ExecError, %(Command `#{command}` failed with output: #{output})
         end
         output
+      rescue Errno::ENOENT => e
+        raise ExecError, %(Command `#{command}` failed: #{e.message})
       end
     end
 
     class Archiver
+      def initialize(shell)
+        @shell = shell
+      end
+
+      def create(options)
+      end
+
+      def extension
+      end
+
+      def content_type
+      end
+
+      def metadata
+        {}
+      end
+    end
+
+    class Tarchiver < Archiver
       def create(options)
-        Puck::Jar.new(options).create!
+        FileUtils.mkdir_p('build')
+        entries = Dir['*']
+        entries.delete('build')
+        @shell.exec("tar czf build/#{options[:archive_name]} #{entries.join(' ')}")
+      end
+
+      def extension
+        'tgz'
+      end
+
+      def content_type
+        'application/x-gzip'
+      end
+    end
+
+    if defined? JRUBY_VERSION
+      begin
+        require 'puck'
+
+        class PuckArchiver < Archiver
+          def create(options)
+            options = options.dup
+            options[:jar_name] = options.delete(:archive_name)
+            Puck::Jar.new(options).create!
+          end
+
+          def extension
+            'jar'
+          end
+
+          def content_type
+            'application/java-archive'
+          end
+
+          def metadata
+            jruby_jars_path = $LOAD_PATH.grep(/\/jruby-jars/).first
+            jruby_version = jruby_jars_path && jruby_jars_path.scan(/\/jruby-jars-(.+)\//).flatten.first
+            if jruby_version
+              super.merge('jruby' => jruby_version)
+            else
+              super
+            end
+          end
+        end
+      rescue LoadError
```

Comment:
I'm not a fan of printing warnings about optional things, so if Puck is going to be optional it's going to have to be silently ignored when not present.

But it's a very good point about the `LoadError`. I've changed it to make sure that only the specific error that "puck" couldn't be loaded is caught.

I'm not too worried about the big `begin … rescue`, it's just a half-screen code, and I think that if it grew any more it should be moved to a separate file, and then we can do the conditional load even nicer:

```ruby
begin
  require 'puck'
  require 'jara/puck_archiver'
rescue LoadError => e
  # check that it was expected
end
```